File: 51-n-queens/n-queens.java
class Solution {
    List<String> construct(char[][] b){
        List<String> a=new ArrayList<>();
        int n=b.length;
        for(int i=0;i<n;i++){
            String c="";
            for(int j=0;j<n;j++){
                c=c+b[i][j];
            }
            a.add(c);
        }
        return a;
    }
    boolean isSafe(char[][] b,int r,int c){
        int n=b.length;
        for(int i=r-1;i>=0;i--){
            if(b[i][c]=='Q') return false;
        }
        int i=r-1,j=c-1;
        while(i>=0 && j>=0){
            if(b[i][j]=='Q') return false;
            i=i-1;
            j=j-1;
        }
        i=r-1;j=c+1;
        while(i>=0 && j<n){
            if(b[i][j]=='Q') return false;
            i=i-1;
            j=j+1;
        }
        return true;
    }
    void check(int r,char[][] b,List<List<String>> ans){
        if(r==b.length){
            ans.add(new ArrayList<>(construct(b)));
            return ;
        }
        for(int c=0;c<b.length;c++){
            if(isSafe(b,r,c)){
                b[r][c]='Q';
                check(r+1,b,ans);
                b[r][c]='.';
            }
        }
    }
    public List<List<String>> solveNQueens(int n) {
        char[][] b=new char[n][n];
        for(int i=0;i<n;i++){
            for(int j=0;j<n;j++){
                b[i][j]='.';
            }
        }
        List<List<String>> ans=new ArrayList<>();
        check(0,b,ans);
        return ans;
    }
}

File: 51-n-queens/n-queens.cpp
class Solution {
private:
    bool isOk(int i, int j, vector<string>& board){
        int n = board[0].length();

        // checking coloumn
        for(int k = i - 1; k >= 0; k --){
            if(board[k][j] == 'Q') return false;
        }

        //checking left diagnols
        int l = i - 1;
        int r = j - 1;
        while (l >= 0 && r >= 0){
            if(board[l][r] == 'Q') return false;
            l -= 1;
            r -= 1;
        }

        //checking right dianols
        l = i - 1;
        r = j + 1;
        while (l >= 0 and r < n){
            if(board[l][r] == 'Q') return false;
            l -= 1;
            r += 1;
        }

        return true;
    }
    void check(int r, vector<string>& board, vector<vector<string>>& ans, int n){
        if(r == n){
            ans.push_back(board);
            return ;
        }

        for(int i = 0; i < n; i++){
            if(board[r][i] == '.' && isOk(r, i, board)){
                board[r][i] = 'Q';
                check(r + 1, board, ans, n);
                board[r][i] = '.';
            }
        }

        return ;
    }
public:
    vector<vector<string>> solveNQueens(int n) {

        vector<vector<string>> ans;
        vector<string> board;
        string temp = "";
        for(int i = 0; i < n; i++) temp += '.';
        for(int i = 0; i < n; i++) board.push_back(temp);

        check(0, board, ans, n);
        return ans;    
    }
};